contributor-docs/adrs/adr-021-automated-prop-docs.md

```diff
@@ -0,0 +1,231 @@
+| Adopted  | —     |
+
+## Context
+
+**Draft PR: https://github.com/primer/react/pull/4466**
+
+### Current process
+
+This is the process I've used in the past. Please correct me if I made any mistakes.
+
+- Write props type in TypeScript source
+- Add JSDoc comments to describe each prop (optional - nicer IDE experience)
+- Manually update `{ComponentName}.docs.json` in the format described in the following ([schema](https://github.com/primer/react/blob/main/packages/react/script/components-json/component.schema.json)):
+  - id
+  - name
+  - status
+  - a11yReviewed
+  - stories
+  - props (needs to match the prop's type in the TypeScript source)
+    - name (required)
+    - type (required)
+    - defaultValue
+    - required
+    - deprecated
+    - description
+  - subcomponents
+  - passthrough (when a component passes through any additional props to an underlying element)
+    - element
+    - url
+- Use the data from each `{ComponentName}.docs.json` to generate a `components.json` file with data we can use for generating docs ([example components.json](https://unpkg.com/browse/@primer/react@36.24.0/generated/components.json))
+
+### Proposed process
+
+- Write props type in TypeScript source
+- Add JSDoc comments to describe each prop
+- Add JSDoc comments with metadata for each component
+  - component ID or parent component ID (if applicable)
+  - status
+  - a11yReviewed
+- Use Docgen to parse component source code and generate a `components.json` file with data we can use for generating docs ([example components.json](https://unpkg.com/browse/@primer/react@36.24.0/generated/components.json))
+
+This will take some work upfront, and we'll need to keep some ["gotchas"](#gotchas) in mind going forward.
+
+#### Suggested metadata documented with JSDoc tags:
+
+- `@alias`: May be used when the name we use to refer to the component is different from the `displayName` (for example: `Item` being named `ActionList.Item`)
+- `@deprecated`: Marks a component as deprecated
+- `@primerid`: The unique ID of the component
+- `@primerstatus`: The [lifecycle status](https://primer.style/guides/component-lifecycle) of the component
+- `@primera11yreviewed`: Whether it has passed a11y design and a11y engineering reviews. If this tag is not set, it defaults to `false`
+- `@primerparentid`: The ID of the parent component that this subcomponent belongs to. Only relevant for "child" components (for example: `ActionList.Item`).
+- `@primerstories`: An optional list of Storybook story IDs to override the story IDs we dynamically calculate in the build script by using the component name, status, and file path.
+
+
+#### Example component
+
+The following code style has given me the best results with Docgen.
+
+```tsx
+// --------------------------------------------------
+// packages/react/src/TestComponent/TestComponent.tsx
+// --------------------------------------------------
+import React, {forwardRef, type FC, type PropsWithChildren} from 'react'
```

Comment:
Yea, I also don't like how fragile this is. Would you be interested in a pairing session where we experiment with a more robust alternative?

Reply:
It's worth noting that the code examples follow the patterns we've been using in newer components such as [Stack](https://github.com/primer/react/tree/main/packages/react/src/Stack).

However, it's not ideal that we have to be careful about how we author because of Docgen.